x/twap/logic_test.go

```diff
@@ -521,3 +522,422 @@ func (s *TestSuite) TestPruneRecords() {
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.NewDec(2),
+				},
+				{
+					baseDenom:  baseRecord.Asset1Denom,
+					quoteDenom: baseRecord.Asset0Denom,
+					overrideSp: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+
+			expectedMostRecentRecords: []expectedResults{
+				{
+					spotPriceA: sdk.NewDec(2),
+					spotPriceB: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+			expectedHistoricalRecords: []expectedResults{
+				// The original record.
+				{
+					spotPriceA: baseRecord.P0LastSpotPrice,
+					spotPriceB: baseRecord.P1LastSpotPrice,
+				},
+				// The new record added.
+				{
+					spotPriceA: sdk.NewDec(2),
+					spotPriceB: sdk.NewDecWithPrec(2, 1),
+				},
+			},
+		},
+		"two-asset; pre-set record at t; updated with spot price error in both denom pairs": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:   baseRecord.Asset0Denom,
+					quoteDenom:  baseRecord.Asset1Denom,
+					overrideErr: spError,
+				},
+				{
+					baseDenom:   baseRecord.Asset1Denom,
+					quoteDenom:  baseRecord.Asset0Denom,
+					overrideSp:  sdk.NewDecWithPrec(2, 1),
+					overrideErr: spError,
+				},
+			},
+
+			expectedMostRecentRecords: []expectedResults{
+				{
+					spotPriceA:    sdk.ZeroDec(),
+					spotPriceB:    sdk.NewDecWithPrec(2, 1),
+					lastErrorTime: baseRecord.Time.Add(time.Second), // equals to block time
+				},
+			},
+			expectedHistoricalRecords: []expectedResults{
+				// The original record.
+				{
+					spotPriceA: baseRecord.P0LastSpotPrice,
+					spotPriceB: baseRecord.P1LastSpotPrice,
+				},
+				// The new record added.
+				{
+					spotPriceA:    sdk.ZeroDec(),
+					spotPriceB:    sdk.NewDecWithPrec(2, 1),
+					lastErrorTime: baseRecord.Time.Add(time.Second), // equals to block time
+				},
+			},
+		},
+		"two-asset; pre-set record at t; large spot price in one of the pairs": {
+			preSetRecords: []types.TwapRecord{baseRecord},
+			poolId:        baseRecord.PoolId,
+			blockTime:     baseRecord.Time.Add(time.Second),
+
+			spOverrides: []spOverride{
+				{
+					baseDenom:  baseRecord.Asset0Denom,
+					quoteDenom: baseRecord.Asset1Denom,
+					overrideSp: sdk.OneDec(),
+				},
+				{
+					baseDenom:   baseRecord.Asset1Denom,
+					quoteDenom:  baseRecord.Asset0Denom,
+					overrideSp:  types.MaxSpotPrice.Add(sdk.OneDec()),
+					overrideErr: nil, // twap logic should identify the large spot price and mark it as error.
+				},
+			},
+
+			expectedMostRecentRecords: []expectedResults{
+				{
+					spotPriceA:    sdk.OneDec(),
+					spotPriceB:    types.MaxSpotPrice,               // Although the price returned from AMM was MaxSpotPrice + 1, it is reset to just MaxSpotPrice.
```

Comment:
oh interesting so we cap at max spot price regardless of having greater value than sp

Reply:
Yes we cap it if it is greater but we set it to zero if we get an error